### PR TITLE
KAFKA-12187: replace assertTrue(obj instanceof X) with assertInstanceOf

### PR DESCRIPTION
**Pull request by brandboat**
as title, replace 
- `assertTrue(obj instanceof X)` with `assertInstanceOf(X.class, obj)`
- `assertTrue(obj instanceof X, errormessge)` with `assertInstanceOf(X.class, obj, errormessage)`
- `assertTrue(obj.isInstance(X))` with `assertInstanceOf(X, obj)`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 1]
```diff
@@ -850,7 +851,7 @@ public void testMetadataTimeoutWithMissingTopic(boolean isIdempotenceEnabled) th
         try {
             future.get();
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Does it expect to get `ExecutionException`? if so, could you please rewrite it by `assertThrow`?

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 2]
```diff
@@ -919,7 +920,7 @@ public void testMetadataTimeoutWithPartitionOutOfRange(boolean isIdempotenceEnab
...
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: ditto

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java [hunk 3]
```diff
@@ -771,7 +772,7 @@ public void testCanRetryWithoutIdempotence() throws Exception {
         try {
             future.get();
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java]: ditto

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java [hunk 4]
```diff
@@ -2543,7 +2544,7 @@ public void testInflightBatchesExpireOnDeliveryTimeout() throws InterruptedExcep
             request.get();
             fail("The expired batch should throw a TimeoutException");
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java]: ditto

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 5]
```diff
@@ -2718,7 +2719,7 @@ public void testDropCommitOnBatchExpiry() throws InterruptedException {
             responseFuture.get();
             fail("Expected to get a TimeoutException since the queued ProducerBatch should have been expired");
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java]: ditto

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 6]
```diff
@@ -2669,15 +2670,15 @@ public void testTransitionToAbortableErrorOnMultipleBatchExpiry() throws Interru
...
         } catch (ExecutionException e) {
-            assertTrue(e.getCause() instanceof  TimeoutException);
+            assertInstanceOf(TimeoutException.class, e.getCause());
         }
 
         try {
             // make sure the produce was expired.
             secondBatchResponse.get();
             fail("Expected to get a TimeoutException since the queued ProducerBatch should have been expired");
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java]: ditto

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 7]
```diff
@@ -2669,15 +2670,15 @@ public void testTransitionToAbortableErrorOnMultipleBatchExpiry() throws Interru
             firstBatchResponse.get();
             fail("Expected to get a TimeoutException since the queued ProducerBatch should have been expired");
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java]: ditto

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 8]
```diff
@@ -2789,7 +2790,7 @@ public void testTransitionToFatalErrorWhenRetriedBatchIsExpired() throws Interru
             responseFuture.get();
             fail("Expected to get a TimeoutException since the queued ProducerBatch should have been expired");
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java]: ditto

File: clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java [hunk 9]
```diff
@@ -1204,7 +1205,7 @@ public void testClientLoginCallbackOverride() throws Exception {
         try {
             createClientConnection(securityProtocol, "invalid");
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java:1206]: ditto

File: clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java [hunk 10]
```diff
@@ -1809,7 +1810,7 @@ public void testValidSaslOauthBearerMechanismWithoutServerTokens() throws Except
             createEchoServer(securityProtocol);
             fail("Server created with invalid login config containing extensions without a token");
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslAuthenticatorTest.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@brandboat thanks for taking over this hard work. please also fix following use case.

1. https://github.com/apache/kafka/blob/trunk/raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java#L971
2. https://github.com/apache/kafka/blob/trunk/raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java#L455
3. https://github.com/apache/kafka/blob/trunk/tools/src/test/java/org/apache/kafka/tools/MetadataQuorumCommandTest.java#L161
4. https://github.com/apache/kafka/blob/trunk/tools/src/test/java/org/apache/kafka/tools/MetadataQuorumCommandTest.java#L168

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java (hunk 1 above)
Comment (brandboat) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Thanks for the detailed review ! :smiley: 

Here's the revised version for a final check before I proceed with applying it.
```java
try {
    assertInstanceOf(TimeoutException.class, assertThrows(ExecutionException.class, future::get).getCause());
} finally {
    producer.close(Duration.ofMillis(0));
}
```

**Comment by brandboat** [comment]
gentle ping @chia7712, there is another case that could be replaced with assertIsntanceOf 
(in IntelliJ Find in files with regex `assertTrue\(\n?.*isInstance`)
https://github.com/apache/kafka/blob/3f666026260472179c17ded967a8d9577107e1f0/clients/src/test/java/org/apache/kafka/test/TestUtils.java#L559
could be replaced with
```
assertInstanceOf(exceptionCauseClass, exception.getCause(),
```
should we include this one too ?

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java (hunk 1 above)
Comment (chia7712) [reply clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: yep, we are on the same page :)

**Comment by chia7712** [comment]
> should we include this one too ?

yep, please

**Comment by chia7712** [comment]
I have pushed a commit to rebase code and fix conflict. will merge it if there is no connected failure.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java [hunk 11]
```diff
@@ -1037,7 +1038,7 @@ public void testHeartbeatRequestWithFencedInstanceIdException() throws Interrupt
             }
             fail("Expected pollHeartbeat to raise fenced instance id exception in 1 second");
         } catch (RuntimeException exception) {
-            assertTrue(exception instanceof FencedInstanceIdException);
+            assertInstanceOf(FencedInstanceIdException.class, exception);
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java]: this should be rewrite by `assertThrow`

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/WakeupTriggerTest.java [hunk 12]
```diff
@@ -177,7 +177,7 @@ private void assertWakeupExceptionIsThrown(final CompletableFuture<?> future) {
         try {
             future.get(DEFAULT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/WakeupTriggerTest.java]: ditto

File: clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java [hunk 13]
```diff
@@ -2172,7 +2173,7 @@ public void testCreatePartitions() throws Exception {
                 otherTopicResult.get();
                 fail("get() should throw ExecutionException");
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@brandboat could you please also rewrite the `fail` by `assertThrow` for the files which are already changed by this PR?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java (hunk 11 above)
Comment (brandboat) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinatorTest.java]: Huge thanks for the detailed view again ! :smiley: 

> I have pushed a commit to rebase code and fix conflict. will merge it if there is no connected failure.

You are so kind! Next time I'll try to fix the conflict faster...

**Comment by chia7712** [comment]
the failed tests pass on my local:
```sh
./gradlew cleanTest :streams:test --tests HighAvailabilityTaskAssignorIntegrationTest.shouldScaleOutWithWarmupTasksAndPersistentStores :storage:test --tests TransactionsWithTieredStoreTest.testCommitTransactionTimeout :connect:runtime:test --tests org.apache.kafka.connect.integration.OffsetsApiIntegrationTest.testAlterSinkConnectorOffsetsOverriddenConsumerGroupId :metadata:test --tests QuorumControllerTest.testFenceMultipleBrokers --tests QuorumControllerTest.testBalancePartitionLeaders :trogdor:test --tests CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated :connect:mirror:test --tests IdentityReplicationIntegrationTest.testReplicateFromLatest --tests MirrorConnectorsIntegrationBaseTest.testSyncTopicConfigs :core:test --tests PlaintextAdminIntegrationTest.testReplicaCanFetchFromLogStartOffsetAfterDeleteRecords --tests ConsumerBounceTest.testConsumptionWithBrokerFailures --tests ReplicaManagerTest.testRemoteFetchExpiresPerSecMetric --tests PlaintextConsumerTest.testFetchOutOfRangeOffsetResetConfigLatest --tests LogDirFailureTest.testIOExceptionDuringLogRoll --tests LogDirFailureTest.testIOExceptionDuringCheckpoint :clients:test --tests SslTransportLayerTest.testServerTruststoreDynamicUpdate
```
will commit it

**Pull request merged**